File: print_designer/print_designer/page/get_original_vat/get_original_vat.py
# Copyright (c) 2026, AWS Solution Ltd. and contributors
# For license information, please see license.txt

import frappe


def get_context(context):
    context.brand_html = "Get Original VAT"
    context.title = "Get Original VAT"


def flt(v):
    try:
        return float(v or 0)
    except (ValueError, TypeError):
        return 0.0


@frappe.whitelist()
def get_all_vat_transactions():
    """
    Return all VAT transactions by Import Clearance Item rows.
    Each row = one VAT record from the PI's Customs Entries (tabImport Clearance Item).
    """
    if not frappe.has_permission("Purchase Invoice", "read"):
        frappe.throw(frappe._("No permission to read Purchase Invoice"))

    # Query Import Clearance Item rows that have VAT
    # parenttype='Purchase Invoice' means this is a PI customs entry row
    rows = frappe.db.sql(
        """
        SELECT
            ici.name,
            ici.parent,
            ici.idx,
            ici.vendor_name,
            ici.vendor_display_name,
            ici.vat_type,
            ici.base_amount,
            ici.vat_amount,
            ici.is_vat,
            ici.tax_amount,
            ici.description,
            COALESCE(s.supplier_name, ici.vendor_name) AS vendor_display_name,
            pi.supplier_name AS pi_supplier_name,
            pi.supplier AS pi_supplier,
            pi.posting_date,
            pi.docstatus,
            pi.tbs_custom_po_type
        FROM `tabImport Clearance Item` ici
        INNER JOIN `tabPurchase Invoice` pi ON pi.name = ici.parent
        LEFT JOIN `tabSupplier` s ON s.name = ici.vendor_name
        WHERE ici.parenttype = 'Purchase Invoice'
          AND ici.vat_amount > 0
        ORDER BY pi.posting_date DESC, ici.idx ASC
        LIMIT 500
        """,
        as_dict=True,
    )

    transactions = []
    for r in rows:
        # Determine the display vendor name
        vendor = r.vendor_display_name or r.pi_supplier_name or r.pi_supplier or r.vendor_name or "-"

        transactions.append({
            # Core identifiers — name+idx uniquely identify an ICI row
            "name": r.name,
            "row_idx": r.idx,
            # PI link
            "pi_name": r.parent,
            "pi_posting_date": r.posting_date,
            "pi_docstatus": r.docstatus,
            # Vendor / party
            "vendor": vendor,
            "description": r.description or "",
            # VAT details
            "vat_type": r.vat_type or "No VAT",
            "base_amount": flt(r.base_amount),
            "vat_amount": flt(r.vat_amount),
            # Status — will be populated below
            "has_original": 0,
            "is_input_vat_used": 0,
            "tpv_name": "",
            # Display type
            "doctype": "Import Clearance Item",
        })

    # Look up TPV records and mark status
    pi_names = list({t["pi_name"] for t in transactions})
    if pi_names:
        tpvs = frappe.db.sql(
            """
            SELECT tpv.name, tpv.purchase_invoice, tpv.has_original,
                   tpv.is_input_vat_used, tpv.vat_type
            FROM `tabThai Purchase VAT` tpv
            WHERE tpv.purchase_invoice IN %(pi_names)s
            """,
            {"pi_names": pi_names},
            as_dict=True,
        )
        tpv_map = {t.purchase_invoice: t for t in tpvs}

        for t in transactions:
            tpv = tpv_map.get(t["pi_name"])
            if tpv:
                t["has_original"] = tpv.has_original or 0
                t["is_input_vat_used"] = tpv.is_input_vat_used or 0
                t["tpv_name"] = tpv.name

    # Also include IUV records
    iuvs = frappe.db.sql(
        """
        SELECT name, purchase_invoice, supplier, supplier_name,
               posting_date, base_amount, vat_amount,
               source_doctype AS vat_type,
               has_original, docstatus
        FROM `tabInput VAT Undue`
        WHERE docstatus = 1
        ORDER BY posting_date DESC
        LIMIT 100
        """,
        as_dict=True,
    )
    for i in iuvs:
        transactions.append({
            "name": i.name,
            "row_idx": 0,
            "pi_name": i.purchase_invoice,
            "pi_posting_date": i.posting_date,
            "pi_docstatus": i.docstatus,
            "vendor": i.supplier_name or i.supplier or "-",
            "description": "",
            "vat_type": i.vat_type or "-",
            "base_amount": flt(i.base_amount),
            "vat_amount": flt(i.vat_amount),
            "has_original": i.has_original or 0,
            "is_input_vat_used": 0,
            "tpv_name": "",
            "doctype": "Input VAT Undue",
        })

    # Sort by posting date desc
    transactions.sort(key=lambda t: t.get("pi_posting_date") or "", reverse=True)

    # Counts
    draft_count = sum(1 for t in transactions if t["pi_docstatus"] == 0)
    pending_reconcile = sum(1 for t in transactions if t["pi_docstatus"] == 1 and not t["has_original"])
    pending_use = sum(1 for t in transactions if t["has_original"] and not t["is_input_vat_used"])

    return {
        "transactions": transactions,
        "total": len(transactions),
        "pending_pi_draft": draft_count,
        "pending_pi_reconcile": pending_reconcile,
        "pending_pi_use": pending_use,
    }


@frappe.whitelist()
def toggle_reconciled(names, doctype=None):
    """
    Toggle has_original on TPV record for given PI names.
    Creates TPV if it doesn't exist.
    """
    if isinstance(names, str):
        import json
        names = json.loads(names)

    results = []
    for pi_name in names:
        try:
            # Get PI details
            pi = frappe.get_doc("Purchase Invoice", pi_name)
            pd = pi.posting_date

            tpv_name = frappe.db.exists("Thai Purchase VAT", {"purchase_invoice": pi_name})
            if tpv_name:
                frappe.db.set_value("Thai Purchase VAT", tpv_name, "has_original", 1)
            else:
                # Create new TPV via DB insert
                total_vat = flt(getattr(pi, "tbs_import_vat_amount", 0) or 0) + \
                            flt(getattr(pi, "tbs_local_vat_amount", 0) or 0) + \
                            flt(getattr(pi, "tbs_forwarder_vat_amount", 0) or 0) + \
                            flt(getattr(pi, "tbs_3rd_party_vat_amount", 0) or 0)
                po_type = getattr(pi, "tbs_custom_po_type", "General") or "General"
                vat_type = "Import" if po_type == "Import" else ("Undue" if po_type == "Service" else "General")
                tpv_name_new = "THVAT-PI-" + pi_name
                frappe.db.sql(
                    """
                    INSERT INTO `tabThai Purchase VAT` (
                        name, purchase_invoice, purchase_for_company, supplier, supplier_name,
                        posting_date, vat_month, vat_year, vat_type,
                        base_amount, vat_amount, total_amount,
                        has_original, workflow_state, docstatus,
                        owner, modified_by
                    ) VALUES (
                        %(name)s, %(purchase_invoice)s, %(company)s, %(supplier)s, %(supplier_name)s,
                        %(posting_date)s, %(vat_month)s, %(vat_year)s, %(vat_type)s,
                        0, %(vat_amount)s, %(vat_amount)s,
                        1, 'Completed', 0,
                        %(owner)s, %(owner)s
                    )
                    """,
                    {
                        "name": tpv_name_new,
                        "purchase_invoice": pi_name,
                        "company": pi.company,
                        "supplier": pi.supplier,
                        "supplier_name": pi.supplier_name,
                        "posting_date": pd,
                        "vat_month": str(pd.month) if pd else None,
                        "vat_year": str(pd.year) if pd else None,
                        "vat_type": vat_type,
                        "vat_amount": total_vat,
                        "owner": frappe.session.user,
                    },
                )
            results.append({"name": pi_name, "success": True, "tpv_name": tpv_name or tpv_name_new})
        except Exception as e:
            results.append({"name": pi_name, "success": False, "error": str(e)})

    frappe.db.commit()
    return results


@frappe.whitelist()
def toggle_vat_used(names, doctype=None):
    """
    Mark PI (via TPV) or TPV as used in VAT declaration.
    """
    if isinstance(names, str):
        import json
        names = json.loads(names)

    results = []
    for name in names:
        try:
            pi_exists = frappe.db.exists("Purchase Invoice", name)
            if pi_exists:
                # First ensure reconciled (TPV exists)
                toggle_reconciled([name])
                # Then mark TPV as used
                tpv_name = frappe.db.exists("Thai Purchase VAT", {"purchase_invoice": name})
                if tpv_name:
                    pi = frappe.get_doc("Purchase Invoice", name)
                    pd = pi.posting_date
                    frappe.db.set_value("Thai Purchase VAT", tpv_name, "is_input_vat_used", 1)
                    frappe.db.set_value("Thai Purchase VAT", tpv_name, "input_vat_used_month", str(pd.month))
                    frappe.db.set_value("Thai Purchase VAT", tpv_name, "input_vat_used_year", str(pd.year))
                results.append({"name": name, "success": True})
            else:
                # TPV — mark directly
                frappe.db.set_value("Thai Purchase VAT", name, "is_input_vat_used", 1)
                results.append({"name": name, "success": True})
        except Exception as e:
            results.append({"name": name, "success": False, "error": str(e)})

    frappe.db.commit()
    return results


@frappe.whitelist()
def delete_transaction(doctype, name):
    """Delete a VAT transaction record."""
    if doctype == "Input VAT Undue":
        if not frappe.has_permission("Input VAT Undue", "delete"):
            frappe.throw(frappe._("No permission to delete Input VAT Undue"))
        frappe.delete_doc("Input VAT Undue", name, ignore_permissions=True)
        frappe.db.commit()
        return {"success": True, "name": name}
    return {"success": False, "error": "Cannot delete this record type"}

File: print_designer/print_designer/page/get_original_vat/get_original_vat.js
// Copyright (c) 2026, AWS Solution Ltd. and contributors
// For license information, please see license.txt

frappe.pages["get-original-vat"].on_page_load = function (wrapper) {
    var $wrapper = $(wrapper);
    load_and_render($wrapper);
};

function load_and_render($wrapper) {
    $wrapper.html('<div class="text-muted text-center p-5"><i class="fa fa-spinner fa-spin fa-2x"></i></div>');
    frappe.call({
        method: "print_designer.print_designer.page.get_original_vat.get_original_vat.get_all_vat_transactions",
        callback: function (r) {
            if (r.message) render_page($wrapper, r.message);
        },
        error: function () {
            $wrapper.html('<div class="p-4">' + __("Failed to load data") + "</div>");
        },
    });
}

function render_page($wrapper, data) {
    var html = '<div class="desk-page p-4">';
    html += '<style>';
    html += '.desk-page .card { background-color: var(--bg-color); border-color: var(--border-color); }';
    html += '.desk-page .card-header { background-color: var(--bg-color); border-bottom-color: var(--border-color); }';
    html += '.desk-page .card-body { background-color: var(--bg-color); }';
    html += '.desk-page .table { color: var(--text-color); }';
    html += '.desk-page th { background-color: var(--bg-color); }';
    html += '.desk-page .text-muted { color: var(--text-muted) !important; }';
    html += '.desk-page .toggle-reconciled, .desk-page .toggle-vat-used { cursor: pointer; }';
    html += '</style>';

    // Summary row
    html += '<div class="row mb-4">';
    html += summary_card(__("Total Records"), data.total || 0);
    html += summary_card(__("PI Draft"), data.pending_pi_draft || 0);
    html += summary_card(__("Pending Reconcile"), data.pending_pi_reconcile || 0);
    html += summary_card(__("Pending VAT Declaration"), data.pending_pi_use || 0);
    html += "</div>";

    // Table
    html += '<div class="card">';
    html += '<div class="card-header d-flex justify-content-between align-items-center">';
    html += '<h5 class="mb-0">' + __("VAT Transactions") + "</h5>";
    html += '<div>';
    html += '<button class="btn btn-sm btn-secondary mr-1" id="bulk-reconciled">' + __("Mark Reconciled") + "</button>";
    html += '<button class="btn btn-sm btn-secondary" id="bulk-vat-used">' + __("Mark VAT Used") + "</button>";
    html += "</div></div>";
    html += '<div class="card-body p-0"><div class="table-responsive">';
    html += '<table class="table table-hover"><thead><tr>';
    html += '<th style="width:40px;"><input type="checkbox" id="select-all"></th>';
    html += "<th>" + __("PI Name") + "</th>";
    html += "<th>" + __("Posting Date") + "</th>";
    html += "<th>" + __("Vendor / Party") + "</th>";
    html += "<th>" + __("Description") + "</th>";
    html += "<th>" + __("VAT Type") + "</th>";
    html += '<th class="text-right">' + __("Base Amount") + "</th>";
    html += '<th class="text-right">' + __("VAT Amount") + "</th>";
    html += '<th class="text-center">' + __("Reconciled") + "</th>";
    html += '<th class="text-center">' + __("Used") + "</th>";
    html += '<th class="text-center">' + __("Actions") + "</th>";
    html += "</tr></thead><tbody>";

    if (data.transactions && data.transactions.length > 0) {
        data.transactions.forEach(function (r) {
            var is_iuv = r.doctype === "Input VAT Undue";
            var is_draft = r.pi_docstatus === 0;
            var is_reconciled = r.has_original === 1;
            var is_used = r.is_input_vat_used === 1;

            // Format numbers
            var fmt_base = r.base_amount
                ? parseFloat(r.base_amount).toLocaleString("en-US", {minimumFractionDigits: 2})
                : "0.00";
            var fmt_vat = r.vat_amount
                ? parseFloat(r.vat_amount).toLocaleString("en-US", {minimumFractionDigits: 2})
                : "0.00";
            var posting_date = r.pi_posting_date ? frappe.format(r.pi_posting_date, "Date") : "-";

            // App link
            var app = is_iuv ? "input-vat-undue" : "purchase-invoice";
            var pi_link = '<a href="/app/' + app + '/' + r.pi_name + '" target="_blank">' + r.pi_name + '</a>';

            // VAT type badge
            var vat_badge_class = "badge-secondary";
            if (r.vat_type === "Customs VAT") vat_badge_class = "badge-primary";
            else if (r.vat_type === "Inclusive VAT") vat_badge_class = "badge-warning";
            else if (r.vat_type === "Exclusive VAT") vat_badge_class = "badge-info";
            var vat_type_badge = '<span class="badge ' + vat_badge_class + '">' + (r.vat_type || "-") + "</span>";

            html += "<tr data-name='" + r.name + "' data-pi='" + r.pi_name + "'>";
            html += '<td><input type="checkbox" class="record-checkbox" data-pi="' + r.pi_name + '"></td>';
            html += "<td>" + pi_link + "</td>";
            html += "<td>" + posting_date + "</td>";
            html += "<td>" + (r.vendor || "-") + "</td>";
            html += "<td>" + (r.description || "-") + "</td>";
            html += "<td>" + vat_type_badge + "</td>";
            html += '<td class="text-right">' + fmt_base + "</td>";
            html += '<td class="text-right"><strong>' + fmt_vat + "</strong></td>";

            // Reconciled toggle
            if (is_reconciled) {
                html += '<td class="text-center"><i class="fa fa-check text-success toggle-reconciled" data-pi="' + r.pi_name + '" title="' + __("Reconciled — click to undo") + '"></i></td>';
            } else {
                html += '<td class="text-center"><input type="checkbox" class="toggle-reconciled" data-pi="' + r.pi_name + '"></td>';
            }

            // Used toggle — same pattern as Reconciled: always a toggle
            if (is_used) {
                html += '<td class="text-center"><i class="fa fa-check text-success toggle-vat-used" data-pi="' + r.pi_name + '" title="' + __("Used — click to undo") + '"></i></td>';
            } else {
                html += '<td class="text-center"><input type="checkbox" class="toggle-vat-used" data-pi="' + r.pi_name + '"></td>';
            }

            // Actions — always 3 icons: eye, pencil, trash
            html += '<td class="text-center">';
            html += '<a href="/app/' + app + '/' + r.pi_name + '" target="_blank" class="btn btn-xs btn-secondary" title="' + __("View") + '"><i class="fa fa-eye"></i></a> ';
            html += '<a href="/app/' + app + '/' + r.pi_name + '?order=Edit" target="_blank" class="btn btn-xs btn-secondary" title="' + __("Edit") + '"><i class="fa fa-pencil"></i></a> ';
            html += '<button class="btn btn-xs btn-secondary btn-delete" data-pi="' + r.pi_name + '" title="' + __("Delete") + '"><i class="fa fa-trash"></i></button>';
            html += "</td></tr>";
        });
    } else {
        html += '<tr><td colspan="11" class="text-center text-muted p-4">' + __("No records found") + "</td></tr>";
    }

    html += "</tbody></table></div></div></div></div>";
    $wrapper.html(html);

    // --- Event bindings ---

    // Select all
    $wrapper.find("#select-all").on("change", function () {
        $wrapper.find(".record-checkbox").prop("checked", $(this).prop("checked"));
    });

    // Reconciled toggle (checkbox or undo icon)
    $wrapper.find(".toggle-reconciled").on("change", function () {
        var pi = $(this).data("pi");
        toggle_reconciled($wrapper, [pi]);
    });
    $wrapper.find(".toggle-reconciled.fa-check").on("click", function () {
        var pi = $(this).data("pi");
        toggle_reconciled($wrapper, [pi], true); // true = undo
    });

    // Used toggle (checkbox or undo icon)
    $wrapper.find(".toggle-vat-used").on("change", function () {
        var pi = $(this).data("pi");
        toggle_vat_used($wrapper, [pi]);
    });
    $wrapper.find(".toggle-vat-used.fa-check").on("click", function () {
        var pi = $(this).data("pi");
        toggle_vat_used($wrapper, [pi], true); // true = undo
    });

    // Delete
    $wrapper.find(".btn-delete").on("click", function () {
        var pi = $(this).data("pi");
        frappe.confirm(__("Delete customs entry for PI {0}?", [pi]), function () {
            delete_transaction($wrapper, pi);
        });
    });

    // Bulk Mark Reconciled
    $wrapper.find("#bulk-reconciled").on("click", function () {
        var names = [];
        $wrapper.find(".record-checkbox:checked").each(function () {
            names.push($(this).data("pi"));
        });
        if (!names.length) { frappe.msgprint(__("Select records first")); return; }
        toggle_reconciled($wrapper, names);
    });

    // Bulk Mark VAT Used
    $wrapper.find("#bulk-vat-used").on("click", function () {
        var names = [];
        $wrapper.find(".record-checkbox:checked").each(function () {
            names.push($(this).data("pi"));
        });
        if (!names.length) { frappe.msgprint(__("Select records first")); return; }
        toggle_vat_used($wrapper, names);
    });
}

function summary_card(label, value) {
    return '<div class="col-md-3"><div class="card"><div class="card-body">' +
        '<div class="text-muted small mb-1">' + label + "</div>" +
        '<div class="text-large font-weight-bold">' + value + "</div>" +
        "</div></div></div>";
}

function toggle_reconciled($wrapper, names, undo) {
    if (undo) {
        // For now just reload — undo would need backend support
        frappe.msgprint(__("Undo not yet implemented"));
        return;
    }
    frappe.call({
        method: "print_designer.print_designer.page.get_original_vat.get_original_vat.toggle_reconciled",
        args: { names: names },
        callback: function (r) {
            if (r.message) {
                var ok = r.message.filter(function (x) { return x.success; }).length;
                frappe.msgprint({ title: __("Done"), message: __("{0} of {1} marked as reconciled", [ok, names.length]) });
            }
            load_and_render($wrapper);
        },
    });
}

function toggle_vat_used($wrapper, names, undo) {
    if (undo) {
        frappe.msgprint(__("Undo not yet implemented"));
        return;
    }
    frappe.call({
        method: "print_designer.print_designer.page.get_original_vat.get_original_vat.toggle_vat_used",
        args: { names: names },
        callback: function (r) {
            if (r.message) {
                var ok = r.message.filter(function (x) { return x.success; }).length;
                frappe.msgprint({ title: __("Done"), message: __("{0} of {1} marked as used", [ok, names.length]) });
            }
            load_and_render($wrapper);
        },
    });
}

function delete_transaction($wrapper, pi_name) {
    frappe.call({
        method: "print_designer.print_designer.page.get_original_vat.get_original_vat.delete_transaction",
        args: { doctype: "Input VAT Undue", name: pi_name },
        callback: function (r) {
            if (r.message && r.message.success) {
                frappe.msgprint({ title: __("Deleted"), message: pi_name + " " + __("deleted") });
                load_and_render($wrapper);
            } else {
                frappe.msgprint({ title: __("Error"), message: r.message && r.message.error || __("Cannot delete this record type") });
            }
        },
    });
}
